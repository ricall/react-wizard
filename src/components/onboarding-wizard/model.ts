import { ComponentType } from 'react';
import { Schema } from 'yup';
import { IconAdvanced, IconArcade, IconPro } from '@/icons';

export type PageProps = { setPage?: (page: number) => void };
export type Step<T> = {
  name: string;
  description: string;
  schema: Schema<Partial<T>>;
  component: ComponentType<PageProps>;
};

type RecurringCost = {
  monthlyCost: number;
  yearlyCost: number;
};

export const plans = ['ARCADE', 'ADVANCED', 'PRO'] as const;
export type Plan = (typeof plans)[number];

export type PlanDetails = RecurringCost & {
  name: string;
  icon: typeof IconAdvanced;
};
export const planDetails: Record<Plan, PlanDetails> = {
  ARCADE: {
    name: 'Arcade',
    icon: IconArcade,
    monthlyCost: 9,
    yearlyCost: 90,
  },
  ADVANCED: {
    name: 'Advanced',
    icon: IconAdvanced,
    monthlyCost: 12,
    yearlyCost: 120,
  },
  PRO: {
    name: 'Pro',
    icon: IconPro,
    monthlyCost: 15,
    yearlyCost: 150,
  },
};

export const addOns = ['ONLINE', 'STORAGE', 'CUSTOMIZE'] as const;
export type AddOn = (typeof addOns)[number];

export type AddOnDetails = RecurringCost & {
  name: string;
  description: string;
};
export const addOnDetails: Record<AddOn, AddOnDetails> = {
  ONLINE: {
    name: 'Online service',
    description: 'Access to multiplayer games',
    monthlyCost: 1,
    yearlyCost: 10,
  },
  STORAGE: {
    name: 'Larger storage',
    description: 'Extra 1TB of cloud save',
    monthlyCost: 2,
    yearlyCost: 20,
  },
  CUSTOMIZE: {
    name: 'Customizable profile',
    description: 'Custom theme on your profile',
    monthlyCost: 2,
    yearlyCost: 20,
  },
};

export const calculatorFor = (yearlyBilling: boolean) => {
  const costFor = ({ yearlyCost, monthlyCost }: RecurringCost) => (yearlyBilling ? yearlyCost : monthlyCost);
  const formatAmount = (amount: number) => `${amount}/${yearlyBilling ? 'yr' : 'mo'}`;
  const formatCostFor = (cost: RecurringCost) => formatAmount(costFor(cost));

  return {
    costFor,
    formatAmount,
    formatCostFor,
  };
};
