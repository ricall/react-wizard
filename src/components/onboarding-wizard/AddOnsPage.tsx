import * as yup from 'yup';
import { AddOn, addOnDetails, addOns, calculatorFor } from '@/components/onboarding-wizard/model.ts';
import { Checkbox } from '@/components/ui/checkbox.tsx';
import { useFormContext } from 'react-hook-form';
import { cn } from '@/lib/utils.ts';
import { CheckedState } from '@radix-ui/react-checkbox';

export const schema = yup
  .object({
    yearlyBilling: yup.boolean().default(false),
    addOns: yup.array().of(yup.string().oneOf(addOns).required()).default([]),
  })
  .required();
type Form = yup.InferType<typeof schema>;

type AddOnButtonProps = {
  addOn: AddOn;
};
const AddOnButton = ({ addOn }: AddOnButtonProps) => {
  const { watch, setValue } = useFormContext<Form>();

  const values: AddOn[] = watch('addOns') ?? [];
  const yearlyBilling = watch('yearlyBilling');
  const checked = values.includes(addOn);
  const { name, description, ...addOnDetail } = addOnDetails[addOn];
  const { formatCostFor } = calculatorFor(yearlyBilling);

  const handleChange = (state: CheckedState) =>
    setValue('addOns', state === true ? [...values, addOn] : values.filter((value) => value !== addOn));

  return (
    <div className={cn('flex flex-row gap-4 items-center w-full px-4 py-3 rounded-xl border', { 'border-blue-800 bg-gray-50': checked })}>
      <Checkbox className="data-[state=checked]:bg-blue-800" checked={checked} onCheckedChange={handleChange} />
      <div className="flex grow flex-col items-start gap-0">
        <div className="text-sm font-bold">{name}</div>
        <div className="text-xs text-secondary">{description}</div>
      </div>
      <div className="text-sm text-blue-800">+${formatCostFor(addOnDetail)}</div>
    </div>
  );
};

export const AddOnsPage = () => {
  return (
    <div className="flex w-full grow flex-col gap-3">
      {addOns.map((addOn) => (
        <AddOnButton key={addOn} addOn={addOn} />
      ))}
    </div>
  );
};
