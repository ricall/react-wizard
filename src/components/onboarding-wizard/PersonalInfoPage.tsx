import { Label } from '@radix-ui/react-label';
import { Input } from '@/components/ui/input.tsx';
import { useFormContext } from 'react-hook-form';

import * as yup from 'yup';
import 'yup-phone-lite';
import { cn } from '@/lib/utils.ts';

export const schema = yup
  .object({
    name: yup.string().required('This field is required'),
    emailAddress: yup.string().email('This field is not a valid email').required('This field is required'),
    phoneNumber: yup.string().phone('in').required('This field is required'),
  })
  .required();
type Form = yup.InferType<typeof schema>;

export const PersonalInfoPage = () => {
  const {
    register,
    formState: { errors },
  } = useFormContext<Form>();

  return (
    <div className="flex flex-col gap-4">
      <div className="flex flex-col gap-1">
        <div className="flex flex-row text-sm">
          <Label className="text-xs">Name</Label>
          {errors.name && <Label className="flex-1 text-right text-xs font-bold text-red-500">{errors.name.message}</Label>}
        </div>
        <Input
          className={cn('w-full rounded border border-secondary px-2 font-bold', {
            'border-red-500 focus-visible:ring-red-500': errors.name,
          })}
          placeholder="e.g. Stephen King"
          {...register('name')}
        />
      </div>

      <div className="flex flex-col gap-1">
        <div className="flex flex-row text-sm">
          <Label className="text-xs">Email Address</Label>
          {errors.emailAddress && <Label className="flex-1 text-right text-xs font-bold text-red-500">{errors.emailAddress.message}</Label>}
        </div>
        <Input
          className={cn('w-full rounded border border-secondary px-2 font-bold', {
            'border-red-500 focus-visible:ring-red-500': errors.emailAddress,
          })}
          placeholder="e.g. stephenking@lorem.com"
          {...register('emailAddress')}
        />
      </div>

      <div className="flex flex-1 flex-col gap-1">
        <div className="flex flex-row text-sm">
          <Label className="text-xs">Phone Number</Label>
          {errors.phoneNumber && <Label className="flex-1 text-right text-xs font-bold text-red-500">{errors.phoneNumber.message}</Label>}
        </div>
        <Input
          className={cn('w-full rounded border border-secondary px-2 font-bold', {
            'border-red-500 focus-visible:ring-red-500': errors.phoneNumber,
          })}
          placeholder="e.g. +1 1 718 222 2222"
          {...register('phoneNumber')}
        />
      </div>
    </div>
  );
};
