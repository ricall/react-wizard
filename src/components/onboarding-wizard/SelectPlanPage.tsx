import { Label } from '@radix-ui/react-label';
import { Controller, useFormContext } from 'react-hook-form';

import * as yup from 'yup';
import 'yup-phone-lite';
import { Switch } from '@/components/ui/switch.tsx';
import { cn } from '@/lib/utils.ts';
import { Plan, planDetails, plans } from '@/components/onboarding-wizard/model.ts';
import { Button } from '@/components/ui/button.tsx';

export const schema = yup
  .object({
    plan: yup.string().oneOf(['ARCADE', 'ADVANCED', 'PRO'], 'Please select a plan').required('Please select a plan'),
    yearlyBilling: yup.boolean().default(false),
  })
  .required();
type Form = yup.InferType<typeof schema>;

type PlanButtonProps = {
  plan: Plan;
  selected: boolean;
  onChange: (value: Plan) => void;
};
export const PlanButton = ({ plan, selected, onChange }: PlanButtonProps) => {
  const { name, icon: Icon, monthlyCost, yearlyCost, yearlyBonus } = planDetails[plan];
  const { watch } = useFormContext<Form>();
  const yearlyBilling = watch('yearlyBilling');
  const cost = yearlyBilling ? `${yearlyCost}/yr` : `${monthlyCost}/mo`;

  return (
    <Button
      type="button"
      variant="outline"
      className={cn('h-auto md:h-[9rem] w-full flex flex-row md:flex-col gap-4 items-start justify-start rounded-xl p-4', {
        'bg-blue-800 text-white hover:text-white hover:bg-blue-800/80': selected,
      })}
      onClick={() => onChange(plan)}
    >
      <Icon className="min-h-9 min-w-9" />
      <div className="flex h-full flex-col items-start">
        <div className="md:grow" />
        <div>{name}</div>
        <div className="text-secondary">${cost}</div>
        {yearlyBilling && <div className={cn('text-xs', selected ? 'text-white' : 'text-blue-800/80')}>{yearlyBonus}</div>}
      </div>
    </Button>
  );
};

export const SelectPlanPage = () => {
  const {
    watch,
    setValue,
    formState: { errors },
  } = useFormContext<Form>();
  const currentPlan = watch('plan');
  const yearlyBilling = watch('yearlyBilling');
  const handleOnPlanButtonChange = (newPlan: Plan) => setValue('plan', newPlan);

  return (
    <div className="flex grow flex-col gap-4">
      <div className="flex flex-col gap-1">
        {errors.plan && <Label className="flex-1 text-right font-bold text-red-500">{errors.plan.message}</Label>}
        <div className="flex flex-col justify-evenly gap-4 md:flex-row">
          {plans.map((plan) => (
            <PlanButton key={plan} plan={plan} selected={plan === currentPlan} onChange={handleOnPlanButtonChange} />
          ))}
        </div>
      </div>

      <div className="flex w-full justify-center bg-gray-100 py-3">
        <div className="flex flex-row gap-6 align-baseline">
          <Label className={cn('text-xs font-bold text-blue-800', { 'text-secondary': yearlyBilling })}>Monthly</Label>
          <Controller
            name="yearlyBilling"
            render={({ field: { value, onChange } }) => (
              <Switch
                className="-mt-0.5 data-[state=checked]:bg-blue-800 data-[state=unchecked]:bg-blue-800"
                checked={value}
                onCheckedChange={onChange}
              />
            )}
          />
          <Label className={cn('text-xs font-bold text-blue-800', { 'text-secondary': !yearlyBilling })}>Yearly</Label>
        </div>
      </div>
    </div>
  );
};
