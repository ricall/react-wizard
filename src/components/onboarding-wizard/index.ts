export { OnboardingWizard } from '@/components/onboarding-wizard/OnboardingWizard';
export type { Model } from '@/components/onboarding-wizard/steps';
