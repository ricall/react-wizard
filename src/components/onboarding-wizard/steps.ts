import { PersonalInfoPage, schema as personalInfoSchema } from '@/components/onboarding-wizard/PersonalInfoPage.tsx';
import { SelectPlanPage, schema as selectPlanSchema } from '@/components/onboarding-wizard/SelectPlanPage.tsx';
import { AddOnsPage, schema as pickAddOnsSchema } from '@/components/onboarding-wizard/AddOnsPage.tsx';

import * as yup from 'yup';
import { SummaryPage } from '@/components/onboarding-wizard/SummaryPage.tsx';
import { Step } from '@/components/onboarding-wizard/model.ts';

export type Model = yup.InferType<typeof personalInfoSchema> &
  yup.InferType<typeof selectPlanSchema> &
  yup.InferType<typeof pickAddOnsSchema>;

export const steps: Step<Model>[] = [
  {
    name: 'Personal info',
    stepName: 'Your Info',
    description: 'Please provide your name, email address and phone number',
    schema: personalInfoSchema,
    component: PersonalInfoPage,
  },
  {
    name: 'Select your plan',
    stepName: 'Select Plan',
    description: 'You have the option of monthly or yearly billing',
    schema: selectPlanSchema,
    component: SelectPlanPage,
  },
  {
    name: 'Pick add-ons',
    stepName: 'Add-Ons',
    description: 'Add-ons help enhance your gaming experience',
    schema: pickAddOnsSchema,
    component: AddOnsPage,
  },
  {
    name: 'Finishing up',
    stepName: 'Summary',
    description: 'Double-check everything looks OK before confirming',
    schema: selectPlanSchema.concat(pickAddOnsSchema),
    component: SummaryPage,
  },
];
