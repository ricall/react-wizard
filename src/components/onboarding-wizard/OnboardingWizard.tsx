import { cn } from '@/lib/utils.ts';
import { Model, steps } from '@/components/onboarding-wizard/steps.ts';
import { useCallback, useState } from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { Button } from '@/components/ui/button.tsx';
import { SubmittedPage } from '@/components/onboarding-wizard/SubmittedPage.tsx';

type StepProps = {
  index: number;
  label: string;
  selected?: boolean;
};
export const Step = ({ index, label, selected }: StepProps) => {
  return (
    <div className="flex flex-row gap-4 text-sm uppercase text-white">
      <div className={cn('border border-white rounded-full w-8 h-8 text-center content-center', { 'bg-cyan-100 text-black': selected })}>
        {index}
      </div>
      <div className="flex flex-col gap-0 text-xs">
        <div className="text-gray-400">Step {index}</div>
        <div>{label}</div>
      </div>
    </div>
  );
};

type Props = {
  onSubmit: (model: Model) => Promise<void>;
};
export const OnboardingWizard = ({ onSubmit }: Props) => {
  const [currentStep, setCurrentStep] = useState(0);
  const [isSubmitted, setIsSubmitted] = useState(false);

  const { name, description, schema, component: StepComponent } = steps[currentStep];
  const methods = useForm<Model>({
    resolver: yupResolver(schema as never),
    mode: 'onBlur',
  });
  const { handleSubmit, trigger } = methods;

  const setPage = useCallback((page: number) => setCurrentStep(page - 1), []);
  const onNext = async () => {
    const valid = await trigger();
    if (valid) {
      setCurrentStep((prev) => prev + 1);
    }
  };
  const onBack = () => setCurrentStep((prev) => prev - 1);
  const onFormSubmitted = async (model: Partial<Model>) => {
    setIsSubmitted(true);

    await onSubmit(model as Model);

    // The form should be closed - for now we just restart it
    methods.reset();
    setCurrentStep(0);
    setIsSubmitted(false);
  };

  return (
    <div className="flex h-screen">
      <div className="m-auto flex min-h-[30rem] w-full max-w-[50rem] flex-row gap-4 rounded-3xl bg-background p-4">
        <div className="flex h-auto w-56 flex-col gap-6 rounded-xl bg-sidebar-desktop bg-cover bg-no-repeat p-8">
          {steps.map(({ name }, index) => (
            <Step key={name} index={index + 1} label={name} selected={index === currentStep} />
          ))}
        </div>

        <FormProvider {...methods}>
          <form onSubmit={handleSubmit(onFormSubmitted)}>
            <div className="flex h-full w-[32rem] flex-col gap-4 px-14 py-8">
              {!isSubmitted && (
                <>
                  <header className="flex flex-col gap-2">
                    <div className="text-3xl font-bold">{name}</div>
                    <div className="text-sm text-secondary">{description}</div>
                  </header>

                  <StepComponent setPage={setPage} />

                  <div className="flex flex-row">
                    {currentStep > 0 && (
                      <Button className="w-20 text-secondary" type="button" variant="ghost" size="lg" onClick={onBack}>
                        Go Back
                      </Button>
                    )}
                    <div className="grow" />
                    {currentStep < steps.length - 1 && (
                      <Button className="w-20" type="button" variant="primary" size="lg" onClick={onNext}>
                        Next Step
                      </Button>
                    )}
                    {currentStep === steps.length - 1 && (
                      <Button className="w-20" type="submit" variant="primary" size="lg">
                        Confirm
                      </Button>
                    )}
                  </div>
                </>
              )}

              {isSubmitted && <SubmittedPage />}
            </div>
          </form>
        </FormProvider>
      </div>
    </div>
  );
};
