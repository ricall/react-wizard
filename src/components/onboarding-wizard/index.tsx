import { cn } from '@/lib/utils.ts';
import { Model, steps } from '@/components/onboarding-wizard/steps.ts';
import { useCallback, useState } from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { Button } from '@/components/ui/button.tsx';
import { SubmittedPage } from '@/components/onboarding-wizard/SubmittedPage.tsx';

type StepProps = {
  index: number;
  label: string;
  selected?: boolean;
};
export const Step = ({ index, label, selected }: StepProps) => {
  return (
    <div className="flex flex-row text-white text-sm gap-4 uppercase">
      <div className={cn('border border-white rounded-full w-8 h-8 text-center content-center', { 'bg-cyan-100 text-black': selected })}>
        {index}
      </div>
      <div className="flex flex-col gap-0 text-xs">
        <div className="text-gray-400">Step {index}</div>
        <div>{label}</div>
      </div>
    </div>
  );
};

type Props = {
  onSubmit: (model: Model) => Promise<void>;
};
export const OnboardingWizard = ({ onSubmit }: Props) => {
  const [currentStep, setCurrentStep] = useState(0);
  const [isSubmitted, setIsSubmitted] = useState(false);

  const { name, description, schema, component: StepComponent } = steps[currentStep];
  const methods = useForm<Model>({
    resolver: yupResolver(schema as never),
    mode: 'onBlur',
  });
  const { handleSubmit, trigger } = methods;

  const setPage = useCallback((page: number) => setCurrentStep(page - 1), []);
  const onNext = async () => {
    const valid = await trigger();
    if (valid) {
      setCurrentStep((prev) => prev + 1);
    }
  };
  const onBack = () => setCurrentStep((prev) => prev - 1);
  const onFormSubmitted = async (model: Partial<Model>) => {
    setIsSubmitted(true);

    await onSubmit(model as Model);

    // The form should be closed - for now we just restart it
    methods.reset();
    setCurrentStep(0);
    setIsSubmitted(false);
  };

  return (
    <div className="flex h-screen">
      <div className="m-auto w-full max-w-[50rem] bg-background p-4 rounded-3xl flex flex-row gap-4 min-h-[30rem]">
        <div className="bg-sidebar-desktop w-[14rem] h-auto rounded-xl bg-no-repeat bg-cover flex flex-col p-8 gap-6">
          {steps.map(({ name }, index) => (
            <Step key={name} index={index + 1} label={name} selected={index === currentStep} />
          ))}
        </div>

        <FormProvider {...methods}>
          <form onSubmit={handleSubmit(onFormSubmitted)}>
            <div className="flex flex-col gap-4 py-8 px-14 h-full w-[32rem]">
              {!isSubmitted && (
                <>
                  <header className="flex flex-col gap-2">
                    <div className="font-bold text-3xl">{name}</div>
                    <div className="text-secondary text-sm">{description}</div>
                  </header>

                  <StepComponent setPage={setPage} />

                  <div className="flex flex-row">
                    {currentStep > 0 && (
                      <Button className="w-[5rem] text-secondary" type="button" variant="ghost" size="lg" onClick={onBack}>
                        Go Back
                      </Button>
                    )}
                    <div className="grow" />
                    {currentStep < steps.length - 1 && (
                      <Button className="w-[5rem]" type="button" variant="primary" size="lg" onClick={onNext}>
                        Next Step
                      </Button>
                    )}
                    {currentStep === steps.length - 1 && (
                      <Button className="w-[5rem]" type="submit" variant="primary" size="lg">
                        Confirm
                      </Button>
                    )}
                  </div>
                </>
              )}

              {isSubmitted && <SubmittedPage />}
            </div>
          </form>
        </FormProvider>
      </div>
    </div>
  );
};
