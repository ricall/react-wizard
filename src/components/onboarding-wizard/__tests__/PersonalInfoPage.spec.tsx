import { render } from '@testing-library/react';
import { PersonalInfoPage, schema } from '@/components/onboarding-wizard/PersonalInfoPage.tsx';
import { FormProvider, useForm } from 'react-hook-form';
import { Model } from '@/components/onboarding-wizard';
import { PropsWithChildren } from 'react';
import { safeExecute } from '@/components/onboarding-wizard/__tests__/utilities.ts';
import { ValidationError } from 'yup';

const UseFormWrapper = ({ children }: PropsWithChildren) => {
  const methods = useForm<Model>();

  return <FormProvider {...methods}>{children}</FormProvider>;
};

const isValidationError = (error: unknown): error is ValidationError => error instanceof ValidationError;

const VALID_NAME = 'test user';
const VALID_EMAIL = 'test@test.com';
const VALID_PHONE_NUMBER = '+1 1 718 222 2222';

describe('PersonalInfoPage', () => {
  const wrapper = () =>
    render(
      <UseFormWrapper>
        <PersonalInfoPage />
      </UseFormWrapper>,
    );

  it('should render as expected', () => {
    const { asFragment, queryByText, queryByPlaceholderText } = wrapper();

    expect(queryByText('Name')).not.toBeNull();
    expect(queryByPlaceholderText('e.g. Stephen King')).not.toBeNull();
    expect(queryByText('Email Address')).not.toBeNull();
    expect(queryByPlaceholderText('e.g. stephenking@lorem.com')).not.toBeNull();
    expect(queryByText('Phone Number')).not.toBeNull();
    expect(queryByPlaceholderText('e.g. +1 1 718 222 2222')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  describe('schema', () => {
    it('should error when no data is provided', async () => {
      const [error, value] = await safeExecute(
        schema.validate(
          {
            name: '',
            emailAddress: '',
            phoneNumber: '',
          },
          { abortEarly: false },
        ),
      );

      expect(value).toBeUndefined();
      if (!isValidationError(error)) throw new Error('unknown error');

      expect(error.errors).toHaveLength(3);
      expect(error.inner[0]).toMatchObject({ path: 'name', errors: ['This field is required'] });
      expect(error.inner[1]).toMatchObject({ path: 'emailAddress', errors: ['This field is required'] });
      expect(error.inner[2]).toMatchObject({ path: 'phoneNumber', errors: ['This field is required'] });
    });

    it('should validate when valid values are provided', async () => {
      const object = {
        name: VALID_NAME,
        emailAddress: VALID_EMAIL,
        phoneNumber: VALID_PHONE_NUMBER,
      };
      const [error, value] = await safeExecute(schema.validate(object, { abortEarly: false }));

      expect(value).toStrictEqual(object);
      expect(error).toBeUndefined();
    });

    it('should error when user name is not provided', async () => {
      const [error, value] = await safeExecute(
        schema.validate(
          {
            name: '',
            emailAddress: VALID_EMAIL,
            phoneNumber: VALID_PHONE_NUMBER,
          },
          { abortEarly: false },
        ),
      );

      expect(value).toBeUndefined();
      if (!isValidationError(error)) throw new Error('unknown error');

      expect(error.errors).toHaveLength(1);
      expect(error.inner[0]).toMatchObject({ path: 'name', errors: ['This field is required'] });
    });

    it('should error when an invalid email is entered', async () => {
      const [error, value] = await safeExecute(
        schema.validate(
          {
            name: VALID_NAME,
            emailAddress: 'test',
            phoneNumber: VALID_PHONE_NUMBER,
          },
          { abortEarly: false },
        ),
      );

      expect(value).toBeUndefined();
      if (!isValidationError(error)) throw new Error('unknown error');

      expect(error.errors).toHaveLength(1);
      expect(error.inner[0]).toMatchObject({ path: 'emailAddress', errors: ['This field is not a valid email'] });
    });

    it('should error when an invalid phone number is entered', async () => {
      const [error, value] = await safeExecute(
        schema.validate(
          {
            name: VALID_NAME,
            emailAddress: VALID_EMAIL,
            phoneNumber: '+61 123 456',
          },
          { abortEarly: false },
        ),
      );

      expect(value).toBeUndefined();
      if (!isValidationError(error)) throw new Error('unknown error');

      expect(error.errors).toHaveLength(1);
      expect(error.inner[0]).toMatchObject({ path: 'phoneNumber', errors: ['This field is not a valid phone number'] });
    });
  });
});
