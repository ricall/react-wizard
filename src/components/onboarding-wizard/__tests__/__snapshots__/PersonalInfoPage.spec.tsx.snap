// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PersonalInfoPage > should render as expected 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4"
  >
    <div
      class="flex flex-col gap-1"
    >
      <div
        class="flex flex-row text-sm"
      >
        <label
          class="text-xs"
        >
          Name
        </label>
      </div>
      <input
        class="flex h-9 bg-transparent py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 w-full rounded border border-secondary px-2 font-bold"
        name="name"
        placeholder="e.g. Stephen King"
      />
    </div>
    <div
      class="flex flex-col gap-1"
    >
      <div
        class="flex flex-row text-sm"
      >
        <label
          class="text-xs"
        >
          Email Address
        </label>
      </div>
      <input
        class="flex h-9 bg-transparent py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 w-full rounded border border-secondary px-2 font-bold"
        name="emailAddress"
        placeholder="e.g. stephenking@lorem.com"
      />
    </div>
    <div
      class="flex flex-1 flex-col gap-1"
    >
      <div
        class="flex flex-row text-sm"
      >
        <label
          class="text-xs"
        >
          Phone Number
        </label>
      </div>
      <input
        class="flex h-9 bg-transparent py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 w-full rounded border border-secondary px-2 font-bold"
        name="phoneNumber"
        placeholder="e.g. +1 1 718 222 2222"
      />
    </div>
  </div>
</DocumentFragment>
`;
